### DISCUSSION TITLE
Longhorn components & were replicas can be stored

### DISCUSSION THREAD

**Discussion by Hr46ph** [longhorn/longhorn#9686] [Q&A]
I don't know what I did exactly that got me into a pickle, here's a short background version:

- A 5 node baremetal Talos Linux cluster with 3 control planes and 2 worker nodes; I allowed CP's to run workloads. Longhorn could store replicas on any node.
- I got a new worker node, added to the cluster and I changed the setting to not allow CP's to run workloads. I mucked around with eviction and draining nodes to get all the replicas off of the CP's and after some time, that was the case.
- I realized I created a problem. When cordoning and draining a worker node, replicas have nowhere to go. That wasnt a smart move ... so
- I wanted to allow control planes to store replicas. I already force data locality so workloads should always have a local replica. But with CP's able to store replicas I can safely take a worker offline for maintenance. PS. Only storing replicas, not running workloads on CP's.

With that last point I was puzzled how to configure Longhorn. I know about the taint tolerations and node selectors, but no matter what I tried in those options (and restarting longhorn manager pods), the cp's would not start any longhorn components and without those, replicas cannot be stored.

What I ended up doing is editing the daemonsets and default-configmap to add to spec.template.spec:
```
      tolerations:
      - effect: NoSchedule
        key: node-role.kubernetes.io/control-plane
        operator: Exists
```
I did that for longhorn-manager, engine-image-ei-xxx, longhorn-csi-plugin.

I also dit kubectl edit settings.longhorn.io taint-toleration -n longhorn-system and added the toleration but that didn't stick.

I am pretty sure this is not the right way of doing things, but I had maintenance to do and I saw no other way. Also, some components are still not running on control plane nodes:

```
kubectl get pods -o wide --sort-by=.spec.nodeName
NAME                                                READY   STATUS      RESTARTS      AGE     IP             NODE   NOMINATED NODE   READINESS GATES
instance-manager-040c9930c5280b696c61ad7f4cd7cef0   1/1     Running     0             20h     10.244.5.111   cp01   <none>           <none>
engine-image-ei-f4f7aa25-zk67f                      1/1     Running     0             20h     10.244.5.117   cp01   <none>           <none>
longhorn-manager-hl5mv                              2/2     Running     0             19h     10.244.5.61    cp01   <none>           <none>
longhorn-csi-plugin-dlv9p                           3/3     Running     0             20h     10.244.5.86    cp01   <none>           <none>
instance-manager-4d982560b0cb38b261ba2031f31c4b13   1/1     Running     0             20h     10.244.0.93    cp02   <none>           <none>
longhorn-manager-mrvqk                              2/2     Running     0             19h     10.244.0.210   cp02   <none>           <none>
engine-image-ei-f4f7aa25-xrm8r                      1/1     Running     0             20h     10.244.0.82    cp02   <none>           <none>
longhorn-csi-plugin-r22qh                           3/3     Running     0             20h     10.244.0.38    cp02   <none>           <none>
instance-manager-6d551c5133ff68fa052c17812dbce608   1/1     Running     0             20h     10.244.6.124   cp03   <none>           <none>
engine-image-ei-f4f7aa25-lqtjg                      1/1     Running     0             20h     10.244.6.163   cp03   <none>           <none>
longhorn-manager-76tvk                              2/2     Running     1 (19h ago)   19h     10.244.6.140   cp03   <none>           <none>
longhorn-csi-plugin-gzgc4                           3/3     Running     0             20h     10.244.6.110   cp03   <none>           <none>
daily-backup-28821651-xmskm                         0/1     Completed   0             36h     10.244.3.224   wk01   <none>           <none>
daily-trim-28821476-wr4ps                           0/1     Completed   0             39h     10.244.3.252   wk01   <none>           <none>
csi-snapshotter-68b686dc4-kjcts                     1/1     Running     0             21h     10.244.3.77    wk01   <none>           <none>
csi-snapshotter-68b686dc4-5zcpl                     1/1     Running     0             21h     10.244.3.140   wk01   <none>           <none>
engine-image-ei-f4f7aa25-gsksz                      1/1     Running     0             20h     10.244.3.93    wk01   <none>           <none>
csi-resizer-6d5c898684-7xzpg                        1/1     Running     0             21h     10.244.3.28    wk01   <none>           <none>
longhorn-csi-plugin-8pcf2                           3/3     Running     0             20h     10.244.3.156   wk01   <none>           <none>
csi-provisioner-58cc84b487-zj92n                    1/1     Running     0             21h     10.244.3.89    wk01   <none>           <none>
longhorn-manager-nvrpx                              2/2     Running     0             19h     10.244.3.104   wk01   <none>           <none>
longhorn-ui-86874c969c-l6bcp                        1/1     Running     0             18h     10.244.3.202   wk01   <none>           <none>
csi-attacher-d7d65647-mn5xw                         1/1     Running     0             19h     10.244.3.8     wk01   <none>           <none>
instance-manager-15c6467605b1d49ba56c825ffd1709c3   1/1     Running     0             18h     10.244.3.103   wk01   <none>           <none>
hourly-snapshot-delete-28823693-tp7h2               0/1     Completed   0             133m    10.244.1.249   wk02   <none>           <none>
engine-image-ei-f4f7aa25-smj4f                      1/1     Running     0             20h     10.244.1.154   wk02   <none>           <none>
hourly-snapshot-delete-28823813-cpbnj               0/1     Completed   0             13m     10.244.1.118   wk02   <none>           <none>
hourly-snapshots-28823700-8k7kw                     0/1     Completed   0             126m    10.244.1.84    wk02   <none>           <none>
hourly-snapshots-28823760-vf2jj                     0/1     Completed   0             66m     10.244.1.132   wk02   <none>           <none>
hourly-snapshots-28823820-vv2lw                     0/1     Completed   0             6m58s   10.244.1.81    wk02   <none>           <none>
longhorn-manager-9sjf9                              2/2     Running     0             19h     10.244.1.112   wk02   <none>           <none>
instance-manager-0bb406fb741bbce4f90c973a625dc06c   1/1     Running     0             18h     10.244.1.194   wk02   <none>           <none>
hourly-snapshot-cleanup-28823773-h5xv6              0/1     Completed   0             53m     10.244.1.156   wk02   <none>           <none>
hourly-snapshot-cleanup-28823713-498b9              0/1     Completed   0             113m    10.244.1.76    wk02   <none>           <none>
hourly-snapshot-cleanup-28823653-tkwb4              0/1     Completed   0             173m    10.244.1.37    wk02   <none>           <none>
daily-trim-28822916-rgrrn                           0/1     Completed   0             15h     10.244.1.181   wk02   <none>           <none>
longhorn-csi-plugin-f5kbg                           3/3     Running     0             20h     10.244.1.44    wk02   <none>           <none>
hourly-snapshot-delete-28823753-4zvns               0/1     Completed   0             73m     10.244.1.207   wk02   <none>           <none>
csi-attacher-d7d65647-95l25                         1/1     Running     0             19h     10.244.2.165   wk03   <none>           <none>
longhorn-csi-plugin-4v7g8                           3/3     Running     4 (19h ago)   19h     10.244.2.64    wk03   <none>           <none>
instance-manager-e62ee2f49c212600cda5bfe0cb2ec050   1/1     Running     0             19h     10.244.2.66    wk03   <none>           <none>
csi-snapshotter-68b686dc4-22l6x                     1/1     Running     0             18h     10.244.2.222   wk03   <none>           <none>
longhorn-driver-deployer-5fdf6565d-xfn2m            1/1     Running     1 (19h ago)   19h     10.244.2.5     wk03   <none>           <none>
csi-resizer-6d5c898684-wmvk9                        1/1     Running     0             18h     10.244.2.133   wk03   <none>           <none>
engine-image-ei-f4f7aa25-2grnl                      1/1     Running     0             19h     10.244.2.103   wk03   <none>           <none>
csi-resizer-6d5c898684-dth7g                        1/1     Running     0             18h     10.244.2.37    wk03   <none>           <none>
longhorn-manager-l9jcn                              2/2     Running     0             19h     10.244.2.249   wk03   <none>           <none>
csi-provisioner-58cc84b487-m4fkb                    1/1     Running     0             18h     10.244.2.138   wk03   <none>           <none>
csi-provisioner-58cc84b487-kncmm                    1/1     Running     0             18h     10.244.2.204   wk03   <none>           <none>
longhorn-ui-86874c969c-2nhlf                        1/1     Running     0             19h     10.244.2.106   wk03   <none>           <none>
csi-attacher-d7d65647-psvcm                         1/1     Running     0             18h     10.244.2.192   wk03   <none>           <none>
```

I can undo what I did but that will bring me back to the point where only worker nodes run Longhorn components.

Basically, what I want to accomplish is all nodes running longhorn components required for storing replicas, configured the right way that is supported so I won't get into trouble later on with upgrades or settings reverting because I manually edited daemonsets.

Thanks for helping out!

**Comment by DamiaSan** [comment]
cc @derekbit

**Comment by DamiaSan** [comment]
I think the only way is to put a taint on control plane nodes and a toleration to avoid those nodes on all pods you want only on worker nodes like longhorn-manager, engine-image-ei-xxx, longhorn-csi-plugin, ...

**Reply by Hr46ph** [reply]
Thanks I understand but that wasn't really the question. I was hoping someone could take a look at what I explained, review that list of pods and advise how to correct it.

**Comment by DamiaSan** [comment]
@Hr46ph So what is exactly the question? Sorry but I haven't understood well

**Reply by Hr46ph** [reply]
No worries, appreciate you're taking the time. I'll rephrase.

Simply put, on a 6 node cluster, 3 control planes and 3 workers. I want all nodes to store replicas, but only workers should run workloads. Which longhorn components should run on all nodes, which components on control planes only, and which components  on workers only?

The answer to that question will allow me to see if all components are running where they should.

As I explained I ran workloads on control planes and stored replicas there. When I added an extra worker I changed the config to only run workloads and store replicas on workers. At that point no components ran on the control planes which prevented storing replicas. When I realized I couldnt take workers offline with only 3 workers, I needed to store replicas on control planes as well. Setting the tolerations did not change the fact that the control planes were not running any components. Basically, I set the toleration and Longhorn said it allowed storing a replica on the control planes, but evicting a worker would be pending indefinitely because with no longhorn components running on control planes, there was nothing to facilitate that request.

I ended up using `kubectl edit` on longhorn daemonsets to force them to control planes. A dirty fix so to speak. I think I still need to implement a proper fix and possibly undo my edits to the daeonsets.

**Comment by DamiaSan** [comment]
> Simply put, on a 6 node cluster, 3 control planes and 3 workers. I want all nodes to store replicas, but only workers should run workloads. Which longhorn components should run on all nodes, which components on control planes only, and which components on workers only?

There isn't a "real" response to your question. Surely there are components which runs on every node, as you can read [here](https://longhorn.io/docs/1.7.2/concepts/#11-the-longhorn-manager-and-the-longhorn-engine). Moreover, if a node hosts data replica, it will hosts the pod managing it. So you can't have nodes which store replica without workloads.

**Reply by Hr46ph** [reply]
Perhaps we misunderstand each other.

There is the volume used by a pod and there are 2 replicas. These replicas are copies of the volume and they can be stored on nodes other than the one running the workload (the pod using it). Even the node running the pod might use a volume on another node, if data locality is not enforced.

Either way, its not part of the problem. Your link does answer part of my question so I'll be reading that to see if I can learn which parts of Longhorn should run where.

I'm sorry if I have been unclear and I appreciate your help!

Thanks.
